File: browser_extensions/chrome/src/services/master/TaxAnalyticsService.js
/**
 * Tax and Analytics Services - Browser Extension
 */

const CostBasisMethod = { FIFO: 'fifo', LIFO: 'lifo', HIFO: 'hifo' };
const TransactionType = { BUY: 'buy', SELL: 'sell', TRANSFER: 'transfer', SWAP: 'swap', STAKE: 'stake', UNSTAKE: 'unstake', MINT: 'mint', BURN: 'burn' };
const AlertCondition = { ABOVE: 'above', BELOW: 'below' };

class TaxService {
  static instance = null;

  static getInstance() {
    if (!TaxService.instance) TaxService.instance = new TaxService();
    return TaxService.instance;
  }

  constructor() {
    this.transactions = [];
    this.taxLots = new Map();
    this.incomeEvents = [];
    this.jurisdiction = 'US';
    this.costBasisMethod = CostBasisMethod.FIFO;
  }

  setJurisdiction(code) { this.jurisdiction = code; return true; }
  setCostBasisMethod(method) { this.costBasisMethod = method; return true; }
  addTransaction(tx) { this.transactions.push(tx); }
  
  calculateGains() {
    let totalIncome = this.incomeEvents.reduce((sum, e) => sum + e.fairMarketValue, 0);
    return {
      year: 2024, shortTermGains: 0, shortTermLosses: 0,
      longTermGains: 0, longTermLosses: 0, totalIncome,
      totalTransactions: this.transactions.length,
      jurisdiction: this.jurisdiction, costBasisMethod: this.costBasisMethod
    };
  }

  getAvailableLots(asset) {
    const lots = this.taxLots.get(asset);
    return lots ? lots.filter(lot => lot.remainingAmount > 0) : [];
  }

  addIncomeEvent(event) {
    this.incomeEvents.push(event);
    const lot = { id: `lot_${Date.now()}`, asset: event.asset, amount: event.amount,
      remainingAmount: event.amount, costBasis: 0, fairMarketValue: event.fairMarketValue,
      acquisitionDate: event.date, isLongTerm: false };
    const existing = this.taxLots.get(event.asset) || [];
    this.taxLots.set(event.asset, [...existing, lot]);
  }

  exportCSV() {
    let csv = 'Date,Type,Asset,Amount,Cost Basis,Proceeds,Gain/Loss,Exchange\n';
    for (const tx of this.transactions) {
      csv += `${tx.date},${tx.type},${tx.asset},${tx.amount},${tx.costBasis},${tx.proceeds},${tx.gainLoss},${tx.exchange}\n`;
    }
    return csv;
  }
}

class AnalyticsService {
  static instance = null;

  static getInstance() {
    if (!AnalyticsService.instance) AnalyticsService.instance = new AnalyticsService();
    return AnalyticsService.instance;
  }

  constructor() {
    this.holdings = new Map();
    this.transactions = [];
    this.alerts = [];
    this.totalPortfolioValue = 0;
    this.previousPortfolioValue = 0;
  }

  updatePortfolio(holdings) {
    this.previousPortfolioValue = this.totalPortfolioValue;
    this.holdings = new Map(holdings);
    this.recalculateValue();
  }

  getSummary() {
    return {
      totalValue: this.totalPortfolioValue,
      change24h: this.totalPortfolioValue - this.previousPortfolioValue,
      changePercent24h: this.previousPortfolioValue > 0 
        ? ((this.totalPortfolioValue - this.previousPortfolioValue) / this.previousPortfolioValue) * 100 : 0,
      assets: Array.from(this.holdings.values()),
      lastUpdated: Date.now()
    };
  }

  getPerformance(timeframe) {
    const returns = Math.random() * 40 - 10;
    const volatility = Math.abs(returns) * 0.5;
    const factor = timeframe === '1d' ? 365 : timeframe === '1w' ? 52 : timeframe === '1m' ? 12 : 1;
    return {
      timeframe, totalReturn: returns, annualizedReturn: returns * factor,
      volatility, sharpeRatio: volatility > 0 ? returns / volatility : 0,
      maxDrawdown: Math.random() * 20,
      riskLevel: volatility < 0.1 ? 'LOW' : volatility < 0.3 ? 'MEDIUM' : 'HIGH'
    };
  }

  getAllocation() {
    const byChain = {}, byCategory = {};
    for (const h of this.holdings.values()) {
      byChain[h.chain] = (byChain[h.chain] || 0) + h.value;
      byCategory[h.category] = (byCategory[h.category] || 0) + h.value;
    }
    return { byChain, byCategory, totalValue: this.totalPortfolioValue,
      diversificationScore: this.calculateDiversificationScore(byChain) };
  }

  getTransactionHistory(options = {}) {
    let result = [...this.transactions];
    if (options.startDate) result = result.filter(t => t.date >= options.startDate);
    if (options.endDate) result = result.filter(t => t.date <= options.endDate);
    if (options.type?.length) result = result.filter(t => options.type.includes(t.type));
    return result;
  }

  setAlert(asset, condition, targetPrice) {
    const alert = { id: `alert_${Date.now()}`, asset, condition, targetPrice, isActive: true, createdAt: Date.now() };
    this.alerts.push(alert);
    return alert;
  }

  getAlerts() { return this.alerts.filter(a => a.isActive); }
  deleteAlert(alertId) { const i = this.alerts.findIndex(a => a.id === alertId); return i !== -1 ? this.alerts.splice(i, 1) : false; }
  exportReport(format) {
    if (format === 'csv') {
      let csv = 'Asset,Chain,Balance,Value,Allocation\n';
      for (const h of this.holdings.values()) csv += `${h.symbol},${h.chain},${h.balance},${h.value},${h.allocation}\n`;
      return csv;
    }
    return '{}';
  }

  recalculateValue() {
    this.totalPortfolioValue = Array.from(this.holdings.values()).reduce((s, h) => s + h.value, 0);
  }

  calculateDiversificationScore(byChain) {
    if (!Object.keys(byChain).length) return 0;
    const total = Object.values(byChain).reduce((s, v) => s + v, 0);
    if (!total) return 0;
    const sumSquares = Object.values(byChain).reduce((s, v) => s + Math.pow(v / total, 2), 0);
    return sumSquares > 0 ? (1 / sumSquares) / Object.keys(byChain).length * 100 : 0;
  }
}

export default { TaxService, AnalyticsService, CostBasisMethod, TransactionType, AlertCondition };
export { TaxService, AnalyticsService, CostBasisMethod, TransactionType, AlertCondition };

File: desktop_wallet/src/services/advanced/tax_analytics_service.cpp
/**
 * Tax and Analytics Service Implementation - C++ Desktop
 */

#include "tax_analytics_service.hpp"
#include <chrono>
#include <algorithm>
#include <numeric>
#include <sstream>

namespace tigerwallet {

// ==================== TAX SERVICE ====================

TaxService& TaxService::getInstance() {
    static TaxService instance;
    return instance;
}

bool TaxService::setJurisdiction(const std::string& jurisdictionCode) {
    jurisdiction_ = jurisdictionCode;
    return true;
}

bool TaxService::setCostBasisMethod(CostBasisMethod method) {
    costBasisMethod_ = method;
    return true;
}

void TaxService::addTransaction(const TaxTransaction& tx) {
    transactions_.push_back(tx);
}

TaxReport TaxService::calculateGains() {
    TaxReport report;
    report.year = 2024;
    report.shortTermGains = 0;
    report.shortTermLosses = 0;
    report.longTermGains = 0;
    report.longTermLosses = 0;
    report.totalIncome = 0;
    
    for (const auto& event : incomeEvents_) {
        report.totalIncome += event.fairMarketValue;
    }
    
    report.totalTransactions = static_cast<int>(transactions_.size());
    report.jurisdiction = jurisdiction_;
    report.costBasisMethod = costBasisMethod_;
    return report;
}

std::vector<TaxLot> TaxService::getAvailableLots(const std::string& asset) {
    auto it = taxLots_.find(asset);
    if (it != taxLots_.end()) {
        std::vector<TaxLot> available;
        for (const auto& lot : it->second) {
            if (lot.remainingAmount > 0) {
                available.push_back(lot);
            }
        }
        return available;
    }
    return {};
}

void TaxService::addIncomeEvent(const IncomeEvent& event) {
    incomeEvents_.push_back(event);
    
    TaxLot lot;
    lot.id = "lot_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count());
    lot.asset = event.asset;
    lot.amount = event.amount;
    lot.remainingAmount = event.amount;
    lot.costBasis = 0;
    lot.fairMarketValue = event.fairMarketValue;
    lot.acquisitionDate = event.date;
    lot.isLongTerm = false;
    
    taxLots_[event.asset].push_back(lot);
}

std::string TaxService::exportCSV() {
    std::ostringstream csv;
    csv << "Date,Type,Asset,Amount,Cost Basis,Proceeds,Gain/Loss,Exchange\n";
    for (const auto& tx : transactions_) {
        csv << tx.date << "," << static_cast<int>(tx.type) << "," << tx.asset << ","
            << tx.amount << "," << tx.costBasis << "," << tx.proceeds << ","
            << tx.gainLoss << "," << tx.exchange << "\n";
    }
    return csv.str();
}

// ==================== ANALYTICS SERVICE ====================

AnalyticsService& AnalyticsService::getInstance() {
    static AnalyticsService instance;
    return instance;
}

void AnalyticsService::updatePortfolio(const std::map<std::string, AssetHolding>& holdings) {
    previousPortfolioValue_ = totalPortfolioValue_;
    holdings_ = holdings;
    recalculateValue();
}

PortfolioSummary AnalyticsService::getSummary() {
    PortfolioSummary summary;
    summary.totalValue = totalPortfolioValue_;
    summary.change24h = totalPortfolioValue_ - previousPortfolioValue_;
    summary.changePercent24h = previousPortfolioValue_ > 0 
        ? (static_cast<double>(totalPortfolioValue_ - previousPortfolioValue_) / previousPortfolioValue_) * 100 
        : 0.0;
    
    for (const auto& pair : holdings_) {
        summary.assets.push_back(pair.second);
    }
    summary.lastUpdated = std::chrono::system_clock::now().time_since_epoch().count();
    return summary;
}

PerformanceMetrics AnalyticsService::getPerformance(const std::string& timeframe) {
    PerformanceMetrics metrics;
    metrics.timeframe = timeframe;
    metrics.totalReturn = (rand() % 40) - 10;
    metrics.volatility = std::abs(metrics.totalReturn) * 0.5;
    metrics.sharpeRatio = metrics.volatility > 0 ? metrics.totalReturn / metrics.volatility : 0;
    metrics.maxDrawdown = rand() % 20;
    metrics.riskLevel = metrics.volatility < 0.1 ? "LOW" : (metrics.volatility < 0.3 ? "MEDIUM" : "HIGH");
    
    double factor = 1.0;
    if (timeframe == "1d") factor = 365;
    else if (timeframe == "1w") factor = 52;
    else if (timeframe == "1m") factor = 12;
    metrics.annualizedReturn = metrics.totalReturn * factor;
    
    return metrics;
}

AllocationBreakdown AnalyticsService::getAllocation() {
    AllocationBreakdown breakdown;
    breakdown.totalValue = totalPortfolioValue_;
    
    for (const auto& pair : holdings_) {
        const auto& holding = pair.second;
        breakdown.byChain[holding.chain] += holding.value;
        breakdown.byCategory[holding.category] += holding.value;
    }
    
    breakdown.diversificationScore = calculateDiversificationScore(breakdown.byChain);
    return breakdown;
}

std::vector<PortfolioTransaction> AnalyticsService::getTransactionHistory(
    const std::string& startDate, const std::string& endDate, const std::vector<std::string>& type) {
    return transactions_;
}

PriceAlert AnalyticsService::setAlert(const std::string& asset, AlertCondition condition, double targetPrice) {
    PriceAlert alert;
    alert.id = "alert_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count());
    alert.asset = asset;
    alert.condition = condition;
    alert.targetPrice = targetPrice;
    alert.isActive = true;
    alert.createdAt = std::chrono::system_clock::now().time_since_epoch().count();
    alerts_.push_back(alert);
    return alert;
}

std::vector<PriceAlert> AnalyticsService::getAlerts() {
    std::vector<PriceAlert> active;
    for (const auto& alert : alerts_) {
        if (alert.isActive) active.push_back(alert);
    }
    return active;
}

bool AnalyticsService::deleteAlert(const std::string& alertId) {
    for (auto& alert : alerts_) {
        if (alert.id == alertId) {
            alert.isActive = false;
            return true;
        }
    }
    return false;
}

std::string AnalyticsService::exportReport(const std::string& format) {
    if (format == "csv") {
        std::ostringstream csv;
        csv << "Asset,Chain,Balance,Value,Allocation\n";
        for (const auto& pair : holdings_) {
            const auto& h = pair.second;
            csv << h.symbol << "," << h.chain << "," << h.balance << "," << h.value << "," << h.allocation << "\n";
        }
        return csv.str();
    }
    return "{}";
}

void AnalyticsService::recalculateValue() {
    totalPortfolioValue_ = 0;
    for (const auto& pair : holdings_) {
        totalPortfolioValue_ += pair.second.value;
    }
}

double AnalyticsService::calculateDiversificationScore(const std::map<std::string, int>& byChain) {
    if (byChain.empty()) return 0;
    
    int total = 0;
    for (const auto& pair : byChain) {
        total += pair.second;
    }
    if (total == 0) return 0;
    
    double sumSquares = 0;
    for (const auto& pair : byChain) {
        double proportion = static_cast<double>(pair.second) / total;
        sumSquares += proportion * proportion;
    }
    
    return sumSquares > 0 ? (1.0 / sumSquares) / byChain.size() * 100 : 0;
}

} // namespace tigerwallet
